test/unit/cpu/test-indexsets.cxx

```diff
@@ -37,6 +48,106 @@ TEST_F(IndexSetTest, InvalidSegments)
   EXPECT_NE(true, index_sets_[0].push_back_nocopy(&rs_segment));
 }
 
+TYPED_TEST_P(DependentIndexSetTest, ForwardDependentIndexSets)
+{
+  using ExecPolicy = typename std::tuple_element<0, TypeParam>::type;
+
+  const int numSegments = 10;
+  const int sizeOfSegment = 1;
+  RAJA::IndexSet indexSet;
+  
+  for (int i = 0; i < numSegments; ++i) {
+    indexSet.push_back(RAJA::RangeSegment(i * sizeOfSegment, i * sizeOfSegment + 1));
+  }
+
+  indexSet.initDependencyGraph();
+
+  for (int i = 0; i < numSegments; ++i) {
+    RAJA::DepGraphNode* node = indexSet.getSegmentInfo(i)->getDepGraphNode();
+    if (i < numSegments - 1) {    
+      node->numDepTasks() = 1;
+      node->depTaskNum(0) = i + 1;
+    }
+    if (i > 0) {
+      node->semaphoreValue() = 1;
+    }
+  }
+  indexSet.finalizeDependencyGraph();
+
+
+  std::set<int> visitedNodes;
+  RAJA::forall<ExecPolicy>(indexSet,
+    [&](int i) {
+    for (int j = 0; j < i; ++j) {
+      EXPECT_TRUE(visitedNodes.find(j) != visitedNodes.end()) << "On task: " << i << ", unable to find dependent node: " << j << "\n";
+    }
+    visitedNodes.insert(i);
+  });
+
+}
+
+TYPED_TEST_P(DependentIndexSetTest, BackwardDependentIndexSets)
+{
+  using ExecPolicy = typename std::tuple_element<0, TypeParam>::type;
+
+  const int numSegments = 10;
+  const int sizeOfSegment = 1;
+  RAJA::IndexSet indexSet;
+  
+  for (int i = 0; i < numSegments; ++i) {
+    indexSet.push_back(RAJA::RangeSegment(i * sizeOfSegment, i * sizeOfSegment + 1));
+  }
+
+  indexSet.initDependencyGraph();
+
+  for (int i = 0; i < numSegments; ++i) {
+    RAJA::DepGraphNode* node = indexSet.getSegmentInfo(i)->getDepGraphNode();
+    if (i > 0) {    
+      node->numDepTasks() = 1;
+      node->depTaskNum(0) = i - 1;
+    }
+    if (i < numSegments - 1) {
+      node->semaphoreValue() = 1;
+    }
+  }
+  indexSet.finalizeDependencyGraph();
+
+
+  std::set<int> visitedNodes;
+  RAJA::forall<ExecPolicy>(indexSet,
+    [&](int i) {
+    for (int j = numSegments - 1; j > i; --j) {
+      EXPECT_TRUE(visitedNodes.find(j) != visitedNodes.end()) << "On task: " << i << ", unable to find dependent node: " << j << "\n";
+    }
+    visitedNodes.insert(i);
+  });
+
+}
+
+REGISTER_TYPED_TEST_CASE_P(DependentIndexSetTest, ForwardDependentIndexSets, BackwardDependentIndexSets);
+
+#ifdef RAJA_ENABLE_AGENCY
+
+using agencyExecTypes = ::testing::Types<
+  std::tuple<RAJA::IndexSet::ExecPolicy<RAJA::agency_taskgraph_segit, RAJA::agency_parallel_exec> >,
```

Comment:
We don't really need to use the `std::tuple<>`s here; we can just have the types themselves, which will make it a bit cleaner (and we can just directly use `TypeParam` in the tests).